File: python/2015/005/part_two.py
import common


def is_nice(test):
    found_duplicate_pair = False
    for index in range(0, len(test) - 1):
        one, two = test[index], test[index + 1]
        if one + two in test[index + 2:]:
            found_duplicate_pair = True
            break

    if not found_duplicate_pair:
        return False

    for index in range(0, len(test) - 2):
        one, two = test[index], test[index + 2]
        if one == two:
            return True
        index += 1

    return False


if __name__ == "__main__":
    for input_file in common.inputs:
        nices = 0
        for line in common.read_file(input_file):
            nice = is_nice(line)
            if nice:
                nices += 1

        print(nices)

File: kotlin/src/main/kotlin/pt/dmb/adventofcode/y2015/p05/PartTwo.kt
package pt.dmb.adventofcode.y2015.p05

import pt.dmb.adventofcode.common.Utilities

class PartTwo {
    private var nices = 0

    fun isNice(test: String): Boolean {
        var foundDuplicatePair = false
        for (a in 0 until test.length - 1) {
            val oneTwo = test[a].toString() + test[a + 1].toString()
            if (oneTwo in test.substring(a + 2)) {
                foundDuplicatePair = true
                break
            }
        }

        if (!foundDuplicatePair) return false

        var index = 0
        while (index < test.length - 2) {
            if (test[index] == test[index + 2]) {
                return true
            }
            index++
        }

        return false
    }

    fun increaseNices() {
        nices++
    }

    fun countNices() = nices
}

fun main() {
    val solver = PartTwo()
    Utilities.readInputFile(2015, 5, scope = "main").forEach { line ->
        if (solver.isNice(line)) {
            solver.increaseNices()
        }
    }
    println(solver.countNices())
}